app/core/user/user_service.py

```diff
@@ -1,39 +1,111 @@
+from typing import List, Optional, Dict
+
 import structlog
 
-from app.core.user.model.base import UserUpdate, UserCreate
+from app.core.user.model.base import UserUpdate, UserCreate, UserRead
+from app.core.user.model.skill_plan import SkillPlan, SkillPlanObjective
+from app.core.user.model.user_score import UserScore
 from app.database import get_db_client
 from app.repositories.base import BaseRepository
 from app.repositories.user import UserRepository
 
 log = structlog.get_logger()
 
-user_repo = UserRepository(
-    db_client_factory=get_db_client, db_name="luso", collection_name="users"
-)
 
+class UserService:
+    def __init__(self):
+        self.user_repo = UserRepository(
+            db_client_factory=get_db_client, db_name="luso", collection_name="users"
+        )
+
+    async def create_user(self, user: UserCreate):
+        self._set_ids(user)
+
+        await self.user_repo.create(user)
+
+    async def update_user(self, user_id: str, user: UserUpdate):
+        old_user = await self.user_repo.get(user_id)
+
+        self._set_ids(user)
+        self._remove_scopes(user)
+
+        log.debug(f"updating user points", points=user.score)
+        self._set_user_points(old_user, user)
+        log.debug(f"updating user points", points=user.score)
+
+        await self.user_repo.update(user_id, user)
+
+    def _remove_scopes(self, user: UserUpdate):
+        user.scopes = None
+
+    def _set_ids(self, user):
+        if user.plans is not None:
+            for plan in user.plans:
+                if not plan.id:
+                    log.debug(f"plan id missing for user {user.username}")
+                    plan.id = BaseRepository.generate_uuid()
+                for objective in plan.objectives:
+                    if not objective.id:
+                        objective.id = BaseRepository.generate_uuid()
+
+    def _set_user_points(self, current_user: UserRead, updated_user: UserUpdate):
+        updated_user.score = current_user.score
+
+        current_plan_objective_map = self._generate_plan_objective_map(
+            current_user.plans
+        )
+        updated_plan_objective_map = self._generate_plan_objective_map(
+            updated_user.plans
+        )
 
-async def create_user(user: UserCreate):
-    _set_ids(user)
+        for p_id, p_obj in updated_plan_objective_map.items():
```

Comment:
maybe split this bad boy into smaller methods

Reply:
I think it looks longer because of formatting, there is not good logical separation for this for loop.